Ansible scripts for zos

mhdawson opened the pull request:
Opening for initial comments. Will update commit comment on first revision.

mhdawson commented:
Once question is who to change the remote_user properly as what is in the commit is obviously not right for other platforms.

mhdawson commented:
These are now good enough so that we can run libuv jobs on zOS.  See https://ci.nodejs.org/job/libuv-test-commit-zos/92/

@jbergstroem any suggestions on a better way to change the remote_user ?

jbergstroem commented:
Sorry for the silence. Will review shortly.

mhdawson commented:
@jbergstroem thanks, I think we are about ready to image additional machines and would be great to land before then.

joaocgreis commented:
Some notes:
- The `remote_user` issue must be solved before landing, as landing it in its current form will break every other system. Strange that the override in the section does not work.
- Is there a mechanism to restart Jenkins if it crashes (monit/...)? This can be done later, but please consider it as it's quite important to keep the server reliably online.
- Is `home` needed in the `group_vars` when it is defined in `ansible.cfg`?

Other than that rubber stamp LGTM. I didn't test this but nothing looks wrong.

joaocgreis commented:
@mhdawson does a `user` tag in the inventory also not work? Like this:

https://github.com/nodejs/build/blob/1512c6c1c583e188eb04b1fd880fcc639f8e5d1e/ansible/inventory.yml#L40

mhdawson commented:
@joaocgreis thanks for the comments will rebase and look to see how I can address them.

mhdawson commented:
I've rebased

The suggestion from @joaocgreis allowed me to resolve the remote_user issue with one additional tweak.  Thanks!

I also did some other small clean and addressed an issue I believe would have affected other platforms.

joaocgreis commented on ansible/host_vars/test-marist-zos13-s390x-template:
```diff
@@ -0,0 +1,21 @@
+  RELEASE_DIR: /NODEJS2/python-2017-04-12-py27
+  RELEASE_TYPE: py27
+  GIT_TEMPLATE_DIR: /NODEJS/share/git-core/templates
+server_jobs: 4
```
For this file to exist, add a line with
```
!host_vars/*-template
```
to https://github.com/nodejs/build/blob/master/ansible/.gitignore

gibfahn replied:
Having templates sounds like a great idea in general, so +1 to this.

mhdawson replied:
added

joaocgreis commented on ansible/ansible.cfg:
```diff
@@ -28,3 +28,9 @@ ansible_python_interpreter = /opt/local/bin/python
+ansible_become             = false
+ansible_python_interpreter = /NODEJS2/python-2017-04-12-py27/python27/bin/python
+home                       = /u
+server_jobs: 4
```
This like is in the wrong format. I see it is in the host_vars template, was it forgotten here and should be removed?

joaocgreis commented on ansible/group_vars/test:
```diff
@@ -1,3 +1,5 @@
 jenkins_url: "https://ci.nodejs.org"
 jenkins_worker_jar: "{{ jenkins_url }}/jnlpJars/slave.jar"
 server_user: iojs
+home: "/users"
```
Can this be moved to the host_vars template? Also from `release`.

mhdawson replied:
If I move it out then every time somebody creates as host_vars for other platforms they would have to add that (we currently only have a template for s390x).  It is the default to be used in cases other than z/OS, and its overridden in the ansible.cfg for the z/OS platform.  

I think its better to keep it here than having people need to remember to add it when creating their host_vars files...

joaocgreis replied:
I was reading this wrong before, in that case this is the correct palce. But shouldn't this be `/home` instead?

If I'm reading this right, your objective is to replace references to `/home` with `{{ home }}`, which sounds good to me. But the default should be what was there before replacing, that is `/home` not `/users`. At least the `download slave.jar` step will fail on all other platforms with this line (unless I'm still missing something).

mhdawson replied:
Thanks for catching that, I'm not sure where I came up with users as it should be home.  Pushed commit to fix.

joaocgreis commented on ansible/roles/jenkins-worker/tasks/main.yml:
```diff
@@ -5,22 +5,25 @@
 
 - name: prepare installing tap2junit
-  when: type != "release"
+  when: type != "release" and not os|startswith("zos")
```
Perhaps I'm missing something, but there's a partial added for this for zos. So, is it added but not run for some reason or run somewhere else?

mhdawson replied:
I think I added it hoping it could install tap2junit, but did not manage to get it working.  I've removed the partial for now, and will revisit when we come back to get the parts needed for build Node.js working (right now I've focused on what was needed to build libuv)

gibfahn replied:
Maybe put a `#TODO(mhdawson): get tap2junit working on zOS` for future reference (so we know it's outstanding).

refack commented:
@mhdawson if you need `GYP` I keep my mirror fairly updated https://github.com/refack/gyp

mhdawson commented:
@refack, I've updated to pull from the ci master were we have setup a mechanism to store artifacts.

mhdawson commented:
@joaocgreis pushed changes to address your latest comments. Please take a look.

I've used it to configure the 2 z/OS machines connected to the CI and the libuv job is running nightly (still red though as we have 2 test failures we are working on)

mhdawson commented:
@joaocgreis @gibfahn any chance you can review so that I can get this landed.

joaocgreis commented:
@mhdawson I left a comment inline above, is the only thing blocking this now for me: https://github.com/nodejs/build/pull/796#discussion_r141757097

mhdawson commented:
@joaocgreis pushed commit to fix your last comment.

gibfahn commented:
Given how much zOS specific stuff we now have in `main.yml`, would it be worth spinning that out into a partial? Not sure how easy that is with ansible, and not a blocking request, more a question.

mhdawson commented:
Point taken about trying to pull the zOS stuff out, but would like that to be a future TODO.  The comment I'll add.  Just going to wait to see if @joaocgreis has any other comments to I can do the final pass together if possible.

joaocgreis commented:
@gibfahn I started looking into moving Windows to `ansible/` and I suspect we should probably split `baselayout` into partials like `bootstrap` already is, with a difference: it can have a `default`, so that only `windows` and `aix` need to be exceptions. Still not 100% sure of this though, having groups of subtasks or splitting `baselayout` into smaller roles might make some sense. The best way forward is probably for me (or someone else) to come up with a proposal when time permits.

mhdawson commented:
Landed as 96af6dc55f0baa39db6ef9e8af6e8ae36c56133d
